File: Go-part/Sum.go
package main

import (
	"fmt"
	"runtime"
	"strconv"
	"time"
)

type Sum []int

func (s Sum) Calculate(count, start, end int, flag string, ch chan int) {
	cal := 0

	for i := start; i <= end; i++ {
		for j := 1; j <= 3000000; j++ {
		}
		cal += i
	}

	s[count] = cal
	fmt.Println("flag:", flag, ".")
	ch <- count
}

func (s Sum) LetsGo() {
	const NCPU = 4
	const RANGE = 10000
	var ch = make(chan int)

	runtime.GOMAXPROCS(NCPU)
	for i := 0; i < NCPU; i++ {
		start :=  (RANGE / NCPU) * i + 1
		end := (RANGE / NCPU) * (i + 1)
		go s.Calculate(i, start, end, strconv.Itoa(i + 1), ch)
	}

	for i := 0; i < NCPU; i++ {
		<-ch
	}
}

func main() {
	var s Sum = make([]int, 4, 4)
	var sum int = 0
	var startTime = time.Now()

	s.LetsGo()

	for _, v := range s {
		sum += v
	}

	duration := time.Now().Sub(startTime)

	fmt.Println("Sum is", sum, "; It takes", duration, "seconds.")
}

File: Scala-part/src/main/scala/com/github/b96705008/concurrent/Sum.scala
package com.github.b96705008.concurrent

import akka.actor.{Actor, ActorSystem, Props}
import akka.routing.RoundRobinPool

sealed trait SumTrait
case class Result(value: Int) extends SumTrait

class SumActor extends Actor {
  val RANGE = 10000

  def calculate(start: Int, end: Int, flag: String): Int = {
    var cal = 0

    for (i <- start to end) {
      for (j <- 1 to 3000000) {}
      cal += i
    }

    println(s"flag: $flag.")
    cal
  }

  def receive = {
    case value: Int =>
      val start = (RANGE / 4) * (value - 1) + 1
      val end = (RANGE / 4) * value
      sender ! Result(calculate(start, end, value.toString))
    case _ => println("Unknown in SumActor...")
  }
}

class PrintActor extends Actor {
  def receive = {
    case (sum: Int, startTime: Long) =>
      val duration = (System.nanoTime() - startTime) / 1000000000.0
      println("Sum is " + sum + "; It takes " + duration + " seconds.")
    case _ => println("Unknown in PrintActor...")
  }
}

class MasterActor extends Actor {
  var sum = 0
  var count = 0
  var startTime: Long = 0

  val sumActor = context.actorOf(Props[SumActor]
    .withRouter(RoundRobinPool(nrOfInstances = 4)), name = "sumActor")
  val printActor = context.actorOf(Props[PrintActor], name = "printActor")

  def receive = {
    case "calculate..." =>
      startTime = System.nanoTime()
      for (i <- 1 to 4) sumActor ! i
    case Result(value) =>
      sum += value
      count += 1
      if (count == 4) {
        printActor ! (sum, startTime)
        context.stop(self)
      }
    case _ => println("Unknown in MasterActor...")
  }
}

object Sum {
  def main(args: Array[String]): Unit = {
    var sum = 0

    val system = ActorSystem("MasterActorSystem")
    val masterActor = system.actorOf(Props[MasterActor], name = "masterActor")
    masterActor ! "calculate..."

    Thread.sleep(5000)
    system.terminate()
  }
}
